client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -953,38 +954,77 @@ public final void shutdown() {
+                countDown.countDown();
             } else {
-                LOG.info("closed connection to server [{}:{}]", connectionFactory.getHost(),
-                        connectionFactory.getPort(), closedCon.cause());
+                LOG.error("could not disconnect from the server [{}:{}]", connectionFactory.getHost(), connectionFactory.getPort());
             }
-            connection.disconnect();
-            if (completionHandler != null) {
-                completionHandler.handle(Future.succeededFuture());
+        });
+        try {
+            if (!countDown.await(5, TimeUnit.SECONDS)) {
+                LOG.error("Disconnecting from the server [{}:{}] timed out after 5 seconds", connectionFactory.getHost(), connectionFactory.getPort());
             }
-        }).close();
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect(final Handler<AsyncResult<Void>> completionHandler) {
+        if (disconnecting.compareAndSet(Boolean.FALSE, Boolean.TRUE)) {
+            context.runOnContext(disconnectResult -> {
+                closeConnection(completionHandler);
+            });
+        } else {
+            completionHandler.handle(Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_CONFLICT, "already in the middle of a disconnect operation")));
+        }
     }
+
+    //-----------------------------------< private methods >---
+
+    private void closeConnection(final Handler<AsyncResult<Void>> completionHandler) {
+        if (isConnectedInternal()) {
+            LOG.info("closing connection to server [{}:{}]...", connectionFactory.getHost(), connectionFactory.getPort());
+            connection.disconnectHandler(null); // make sure we are not trying to re-connect
+            connection.closeHandler(closedCon -> {
+                if (closedCon.succeeded()) {
+                    LOG.info("closed connection to server [{}:{}]", connectionFactory.getHost(),
+                            connectionFactory.getPort());
+                } else {
+                    LOG.info("closed connection to server [{}:{}]", connectionFactory.getHost(),
+                            connectionFactory.getPort(), closedCon.cause());
+                }
+                connection.disconnect();
+                disconnecting.compareAndSet(Boolean.TRUE, Boolean.FALSE);
+                if (completionHandler != null) {
+                    completionHandler.handle(Future.succeededFuture());
+                }
+            }).close();
+        } else {
+            LOG.info("connection to server [{}:{}] already closed", connectionFactory.getHost(), connectionFactory.getPort());
+            disconnecting.compareAndSet(Boolean.TRUE, Boolean.FALSE);
+            completionHandler.handle(Future.succeededFuture());
```

Comment:
if completionHandler can indeed be null, then you should check for null here as well ..